### DISCUSSION TITLE
Fix: Imposible to add more domains to whitelist or blocklist when using NFS

### DISCUSSION THREAD

**Discussion by MoJo2600** [MoJo2600/pihole-kubernetes#130] [Q&A] [answered]
The filesystem is readonly, what can I do?

Errors are like this:
> Error, something went wrong!
> While executing INSERT OT IGNORE: attempt to write a readonly database
> Added 0 out of 1 domains

**Comment by MoJo2600** [comment, marked as answer]
### Fixed!
The problem is in the NFS server configuration!

### TL;DR:
Remove `--manage-gids` from the `RPCMOUNTDOPTS` in the file `/etc/default/nfs-kernel-server` on the NFS server and restart the `nfs-kernel-server` service.
 
### Research
So I found out that if user www-data had pihole as the _primary_ group, I was able to write the gravity.db. As soon as I made www-data primary again and pihole the secondary group again, the permission was denied. _However_, if I did the same on the NFS server (by adding group 999 as a secondary group to user www-data), all was fine!

So now I knew it had to do something with the NFS communication. After some DuckDuckGo-ing I found this quote:
[NFS permission problem with secondary groups](https://unix.stackexchange.com/a/206090)
> rpc.mountd(8) - rpc.mountd - NFS mount daemon
> 
> -g or --manage-gids
> Accept requests from the kernel to map user id numbers into lists of group id numbers for use in access control. An NFS request will normally (except when using Kerberos or other cryptographic authentication) contains a user-id and a list of group-ids. Due to a limitation in the NFS protocol, at most 16 groups ids can be listed. **If you use the -g flag, then the list of group ids received from the client _will be replaced_ by a list of group ids determined by an appropriate lookup on the server. Note that _the 'primary' group id is not affected_ so a newgroup command on the client will still be effective.** This function requires a Linux Kernel with version at least 2.6.21.

This means that when `--manage-gids` is passed, the NFS server will try to replace the secondary groups with 'an appropriate lookup on the server'. I'm not 100% sure what that means, but in our situation it had undesired effects.

### Fix
So to fix our permissions problem here, I have edited my `/etc/default/nfs-kernel-server` file on the NFS server and commented the line `RPCMOUNTDOPTS="--manage-gids"`.

Now it works with a fresh install with `values.yml` PVC:
```yaml
persistentVolumeClaim:
  enabled: true
  storageClass: nfs-client
  accessModes:
    - ReadWriteOnce
```
And the default file permissions:
```sh
root@pihole-xxxxxxxxxx-xxxxx:/# ls -lahF /etc/pihole/
total 91M
drwxrwxr-x 3 pihole pihole 4.0K Sep 30 05:50 ./
drwxr-xr-x 1 root   root   4.0K Sep 30 05:46 ../
[...]
-rw-rw-r-- 1 pihole pihole  67M Sep 30 05:47 gravity.db
[...]
```

Hope this helps!

_Note: The config file also recommends to read http://wiki.debian.org/SecuringNFS, so that's a good next step. :-)_

PS: Thanks to @i5Js, your research helped a lot to pinpoint the exact problem!

Thanks to @brnl for this, as discussed here: https://github.com/MoJo2600/pihole-kubernetes/issues/39#issuecomment-701169533
